src/components/organisms/LocalSearch.tsx

```diff
@@ -0,0 +1,979 @@
+import React, {
+  useState,
+  ChangeEventHandler,
+  useCallback,
+  useRef,
+  KeyboardEvent,
+  useEffect,
+} from 'react'
+import { useEffectOnce, useDebounce } from 'react-use'
+import { escapeRegExp } from '../../lib/string'
+import {
+  SEARCH_DEBOUNCE_TIMEOUT,
+  LOCAL_MERGE_SAME_LINE_RESULTS_INTO_ONE,
+  getMatchData,
+} from '../../lib/search/search'
+import CodeMirror, {
+  EditorChangeLinkedList,
+  MarkerRange,
+  TextMarker,
+} from 'codemirror'
+import Icon from '../atoms/Icon'
+import {
+  mdiArrowDown,
+  mdiArrowUp,
+  mdiClose,
+  mdiFormatLetterCase,
+  mdiMagnify,
+  mdiRegex,
+  mdiSubdirectoryArrowLeft,
+} from '@mdi/js'
+import LocalReplace from './LocalReplace'
+import styled from '../../lib/styled/styled'
+import { BaseTheme } from '../../lib/styled/BaseTheme'
+import { SearchReplaceOptions } from './NoteDetail'
+
+const LOCAL_SEARCH_MAX_RESULTS = 10000
+
+interface LocalSearchProps {
+  searchQuery: string
+  replaceQuery: string
+  searchOptions: SearchReplaceOptions
+  codeMirror: CodeMirror.EditorFromTextArea
+  showReplace: boolean
+  onSearchToggle?: (nextState?: boolean) => void
+  onCursorActivity?: (codeMirror: CodeMirror.EditorFromTextArea) => void
+  onSearchQueryChange?: (newSearchQuery: string) => void
+  onReplaceToggle?: (nextState?: boolean) => void
+  onReplaceQueryChange?: (newSearchQuery: string) => void
+  onUpdateSearchOptions: (searchOptions: Partial<SearchReplaceOptions>) => void
+}
+
+export enum SearchResultNavigationDirection {
+  NEXT,
+  PREVIOUS,
+}
+
+// todo: [komediruzecki-13/02/2021] Would this be better as React.Component
+//  are there any benefits from component or this works fine with hooks -
+//  since we need to save state on close (search/Replace value, selected options)?
```

Comment:
Yea, I thought maybe it would help, but later saw that I can do pretty much everything with hooks. 
The initial idea was that since we have some state in the above component, we could transfer that state in this react component (now function component) and then use higher component data to decide how/when to change/bind events and states.

But for now, this works fine even thou the higher (NoteDetail) component is a bit cluttered with event logic and few additional states added.